File: LeetcodePython/3346.py
from bisect import bisect_left, bisect_right
from collections import Counter


class Solution:
    def maxFrequency(self, nums: List[int], k: int, numOperations: int) -> int:
        nums.sort()
        res = 0
        counter = Counter(nums)
        for i in range(nums[0], nums[-1] + 1):
            left = bisect_left(nums, i - k)
            right = bisect_right(nums, i + k)
            res = max(res, min(right - left, counter[i] + numOperations))
        return res

File: LeetcodeCPP/3346.cpp
#include <vector>
#include <algorithm>
#include <unordered_map>
using namespace std;
class Solution {
public:
    int maxFrequency(vector<int>& nums, int k, int numOperations) {
        sort(nums.begin(), nums.end());
        int res = 0;
        unordered_map<int, int> counter;
        
        for (int num : nums) {
            counter[num]++;
        }
        
        for (int i = nums[0]; i <= nums.back(); i++) {
            int left = lower_bound(nums.begin(), nums.end(), i - k) - nums.begin();
            int right = upper_bound(nums.begin(), nums.end(), i + k) - nums.begin();
            int currentCount = counter.count(i) ? counter[i] : 0;
            res = max(res, min(right - left, currentCount + numOperations));
        }
        
        return res;
    }
};